### PR TITLE
[Python] Add support for named parameters

### PR DESCRIPTION
**Pull request by Tishj**
This PR implements named parameters (mentioned in #5388)

After feedback in #5582 , I've taken another stab at implementing this, with a different implementation.
This adds support for named parameters to the parser
In the transform phase they get converted to positional parameters (they also can't be mixed with positional parameters currently, and errors are thrown to make sure of that)
We keep a `case_insensitive_map_t<int32_t>` to remember which name maps to which index, and we move this from the Transformer -> SQLStatement -> PreparedStatement.

Then in the DuckDBPyConnection::Connect() we check this map, and the type of param (same as the earlier PR), and then transform the params to a regular list to be executed as normal

TODO: clear the `named_param_map` when we're done with it, and then check if `named_param_map` is empty in Execute/PendingQuery (?) - Throw error if it's not. This fixes the following situation going under the radar:
```sql
D prepare q123 as select $param, $other_name, $param;
D execute q123 (1, 2);
┌───────┬───────┬───────┐
│  $1   │  $2   │  $1   │
│ int32 │ int32 │ int32 │
├───────┼───────┼───────┤
│     1 │     2 │     1 │
└───────┴───────┴───────┘
```

### CODE REVIEW

**Comment by Tishj** [comment]
@Mause I think with this implementation we won't have issues with the regex example, will verify this

File: src/include/duckdb/parser/transformer.hpp [hunk 1]
```diff
@@ -52,6 +53,8 @@ class Transformer {
 	idx_t max_expression_depth;
 	//! The current prepared statement parameter index
 	idx_t prepared_statement_parameter_index = 0;
+	//! Map from named parameter to parameter index;
+	case_insensitive_map_t<int32_t> named_param_map;
```
Comment (Mytherin) [inline src/include/duckdb/parser/transformer.hpp:57]: `idx_t` maybe?

File: src/parser/transform/expression/transform_param_ref.cpp [hunk 2]
```diff
@@ -10,9 +10,32 @@ unique_ptr<ParsedExpression> Transformer::TransformParamRef(duckdb_libpgquery::P
...
+		D_ASSERT(node->number == 0);
+		int32_t index;
+		if (GetNamedParam(node->name, index)) {
+			// We've seen this named parameter before and assigned it an index!
+			node->number = index;
+		}
+	}
 	if (node->number == 0) {
 		expr->parameter_nr = ParamCount() + 1;
+		if (node->name && !HasNamedParameters() && ParamCount() != 0) {
```
Comment (Mytherin) [inline src/parser/transform/expression/transform_param_ref.cpp:25]: Could we add some sqllogictest tests using named parameters with `PREPARE` and `EXECUTE` (it is fine if they fail - mostly just verifying that it doesn't blow up).

File: src/include/duckdb/parser/sql_statement.hpp [hunk 3]
```diff
@@ -30,6 +31,8 @@ class SQLStatement {
 	idx_t stmt_length = 0;
 	//! The number of prepared statement parameters (if any)
 	idx_t n_param = 0;
+	//! The map of named parameter to param index (if n_param and any named)
+	case_insensitive_map_t<int32_t> named_param_map;
```
Comment (Mytherin) [inline src/include/duckdb/parser/sql_statement.hpp:35]: Could we make this private?

File: tools/pythonpkg/src/pyconnection.cpp [hunk 4]
```diff
@@ -139,6 +140,39 @@ static unique_ptr<QueryResult> CompletePendingQuery(PendingQueryResult &pending_
 	return pending_query.Execute();
 }
 
+py::list TransformNamedParameters(case_insensitive_map_t<int32_t> &named_param_map, const py::dict &params) {
```
Comment (Mytherin) [inline tools/pythonpkg/src/pyconnection.cpp:143]: const reference to the `named_param_map` perhaps?

File: tools/pythonpkg/tests/fast/api/test_duckdb_query.py [hunk 5]
```diff
@@ -25,3 +25,77 @@ def test_duckdb_from_query(self, duckdb_cursor):
...
+                count(*) FILTER (WHERE i >= $1),
+                sum(i) FILTER (WHERE i < $2),
+                avg(i) FILTER (WHERE i < $1)
+            from
+                range(100) tbl(i)
+        """,
+        [5, 10]
+        ).fetchall()
+
+        res = con.execute(
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/api/test_duckdb_query.py:44]: Could we add a test to verify case insensitivity?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. Some comments below:

File: src/parser/transform/expression/transform_param_ref.cpp (hunk 2 above)
Comment (Tishj) [reply src/parser/transform/expression/transform_param_ref.cpp:25]: I've added error detection for this to TransformPrepare, and added a test that shows PREPARE throws when a named parameter is found

Do you think that suffices or does this need more coverage?

Comment (Mytherin) [reply src/parser/transform/expression/transform_param_ref.cpp:25]: I think that's fine

**Comment by Tishj** [comment]
Ah this PR is made against master, so I guess even when it's ready, we should wait for everything be in sync again to merge

**Comment by Mytherin** [comment]
Thanks for the fixes! LGTM. Could you just merge with master again and resolve the merge conflicts?

**Comment by Mytherin** [comment]
Good to go from my side - could you just pull and re-generate the grammar to fix the merge conflicts?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**